### DISCUSSION TITLE
how to http/2 server push static resources

### DISCUSSION THREAD

**Discussion by adrian2x** [fastapi/fastapi#7288] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI, Request
from fastapi.responses import FileResponse
from fastapi.staticfiles import StaticFiles

app = FastAPI()

app.mount("/dist", StaticFiles(directory="views/dist"), name="static")

@app.get("/")
@app.get("/{pathname}")
async def render(pathname: Optional[str] = None, request: Request = None):
    "A catch-all route"
    await request.send_push_promise("/dist/manifest.webmanifest")
    await request.send_push_promise("/dist/index.cbb1cf3e.css")
    await request.send_push_promise("/dist/index.7553f2c4.js")
    return FileResponse("views/dist/index.html")
```

### Description

According to [Starlette docs](https://www.starlette.io/server-push/#requestsend_push_promise), there is a `request.send_push_promise` method for server push. My code runs just fine and serves the route and assets, but they are not server pushed at all, just regular fetch.
How can I get this to work?

### Environment

* OS: Ubuntu 20.04:
* FastAPI Version 0.55.1:
* Python version: 3.7
* Server: hypercorn@0.10.2
* Browser: Chrome 83.x

**Comment by Mause** [comment]
I believe http/2 has certain requirements, such as https? Have you tried over https?

And how do you know they're not being sent over server push?

**Comment by adrian2x** [comment]
@Mause I'm testing the deployed app using this tool https://http2.pro/check, which tells me my server does support http2 but no resources are being pushed even though my main route is configured exactly like above.
Also using Chrome devtools Network tab, pushed resources should show up as `Push / ...` under `initiator` but it's not currently, the initiator is `(index)`

**Comment by v3ss0n** [comment, marked as answer]
you will need hypercorn , have you tried running with it? 
or http2 supported servers.

**Comment by v3ss0n** [comment]
https://pypi.org/project/Hypercorn/

**Comment by adrian2x** [comment]
@v3ss0n I am in fact running hypercorn@0.10.2 but it's not pushing any resources according to Chrome (83)

**Comment by rohanshiva** [comment]
Hey, I am having a similar problem. @adrian2x did you get it to work somehow?

**Comment by Sharaf5** [comment]
@adrian2x, @rohanshiva - I have reached to this write up https://levelup.gitconnected.com/deploy-fastapi-with-hypercorn-http-2-asgi-8cfc304e9e7a?gi=74e947dae0af which may be useful if you miss configured the hypercorn instance

**Comment by adrian2x** [comment]
Thank you for the link, I think I was missing the certificate before and only testing in localhost without encryption.
This is a very nice explanation for anyone planning to run hypercorn.

On Sep 21, 2022, at 11:57 PM, Abdelaziz Sharaf ***@***.******@***.***>> wrote:



@adrian2x<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fadrian2x&data=05%7C01%7C%7Ce9a9d9f72b1443c494a908da9c4e8cff%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637994158448999422%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=%2F1F5pzGZbXpsG3bDNMUw7TvHXodB45MF2s6YKGgXd9M%3D&reserved=0>, @rohanshiva<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frohanshiva&data=05%7C01%7C%7Ce9a9d9f72b1443c494a908da9c4e8cff%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637994158448999422%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=y%2BIwk61bckTky2Bi78P3jgqHxZUAJnCc%2FWNAg0uwiiQ%3D&reserved=0> - I have reached to this write up https://levelup.gitconnected.com/deploy-fastapi-with-hypercorn-http-2-asgi-8cfc304e9e7a?gi=74e947dae0af<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Flevelup.gitconnected.com%2Fdeploy-fastapi-with-hypercorn-http-2-asgi-8cfc304e9e7a%3Fgi%3D74e947dae0af&data=05%7C01%7C%7Ce9a9d9f72b1443c494a908da9c4e8cff%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637994158448999422%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=grD%2Bg9ELjtsCDevDXMo18m7idE4Q9sqeYu3TpSw%2F%2BXo%3D&reserved=0> which may be useful if you miss configured the hypercorn instance

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

If that solves the original problem, then you can close this issue @adrian2x  ✔️
